src/workerd/jsg/promise.h

```diff
@@ -285,6 +285,21 @@ class Promise {
     return kj::mv(promise);
   }
 
+  template <typename... Attachments>
+  Promise<T> attach(Attachments&&... attachments) KJ_WARN_UNUSED_RESULT {
+    auto result = Promise<T>(nullptr);
+    result.v8Promise = kj::mv(v8Promise);
+    result.markedAsHandled = markedAsHandled;
+
+    KJ_IF_SOME(thisAttachments, this->attachments) {
+      result.attachments = kj::mv(thisAttachments).attach(kj::fwd<Attachments>(attachments)...);
+    } else {
+      result.attachments = kj::heap(kj::tuple(kj::fwd<Attachments>(attachments)...));
+    }
+
+    return result;
+  }
```

Comment:
Hmmm... I'd be interested in @kentonv's opinion on this also but I'm not convinced this is the right approach... consider what happens when the `jsg::Promise` is passed across the TypeWrapper system, for instance,

```
jsg::Promise<void> foo(jsg::Lock& js) {
  const promise = doSomething();
  return promise.attach(kj::mv(span));
}
```

The actual `jsg::Promise` itself is going to get moved into the TypeWrapper and eventually dropped on the floor, with the `v8::Promise` it is wrapping outliving both it and the attachments.

If you want the attachments to actually track the completion of the `v8::Promise`, then you really need to attach to it via a continuation (the `promise.whenResolved().then(...)` approach.